src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostMiscTyping.scala

```diff
@@ -213,11 +213,24 @@ trait GhostMiscTyping extends BaseTyping { this: TypeInfoImpl =>
   }
 
   implicit lazy val wellDefSpec: WellDefinedness[PSpecification] = createWellDef {
-    case n@ PFunctionSpec(pres, preserves, posts, terminationMeasures, _, _) =>
+    case n@ PFunctionSpec(pres, preserves, posts, terminationMeasures, privateSpec, _, _) =>
       pres.flatMap(assignableToSpec) ++ preserves.flatMap(assignableToSpec) ++ posts.flatMap(assignableToSpec) ++
       preserves.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode)) ++
       pres.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode)) ++
       terminationMeasures.flatMap(wellDefTerminationMeasure) ++
+      privateSpec.toVector.flatMap(wellDefPrivateSpec) ++
+      // if the function itself is public then specifications outside the private clause cannot be private, 
+      // private members are allowed everywhere if the function itself is private
+      error(n, s"Private members inside a public function must all be defined in the private clause.", {
+        val func = tree.parent(n)
+        val isFuncPvt = func.exists(f => f match {
+          case PFunctionDecl(id, _, _, _, _) => isPrivateReg(id)
+          case PMethodDecl(id, _, _, _, _, _) => isPrivateReg(id)
+          case _ => false
+        })
```

Comment:
I defined a method in the Visibility file:
```
  def getAndcheckIfMemberPrivateFromSpec(spec: PSpecification): (PFunctionOrMethodDecl, Boolean) = {
    val funcOpt = tryEnclosingFunctionOrMethod(spec)
    if (funcOpt.isEmpty) Violation.violation(s"Specification is not enclosed inside a function: $spec")
    val func = funcOpt.get
    (func, isPrivateIdentifier(func.id))
  }
```
So that I can call it in GhostMiscTyping:
`val (func, isFuncPvt) = getAndcheckIfMemberPrivateFromSpec(n)`

Reply:
Rather than creating a method that does this, I now just put:
```
val func = tryEnclosingFunctionOrMethod(n)
val isFuncPvt = if (func.isEmpty) false else isPrivateIdentifier(func.get.id)
```